keps/sig-api-machinery/2309-idl/syntax.md

```diff
@@ -0,0 +1,635 @@
+
+5. Naming in the IDL describes the serialized form.  Language bindings
+   (Go) are derived automatically from that.  Specific overrides for
+   legacy codebases occur via markers, since they are exceptions to the
+   rules.
+
+## Examples
+
+***Author's note***: the cannonical source for these examples is [in the
+repository][repo-example]; the example reproduced here is just to make it
+easier to comment.
+
+[repo-example]: https://github.com/DirectXMan12/idl/blob/main/kdlc/testdata/all-around.kdl
+
+```
+// double slashes add comments
+/* same with block comments */
+
+// Triple-slash comments denote documentation
+// If a line contains `# Words`, the following lines
+// will be placed in the corresponding OpenAPI documentation
+// field.
+
+/// core/v1 describes core k8s concepts
+group-version(group:"core", version:"v1") {
+    /// Pod is a group of related, colocated processes
+    ///
+    /// # Example
+    /// {"apiVersion": "v1", "kind": "Pod", "spec": {}}
+    kind Pod {
+        // types may be nested -- this is just a convenience for namespacing
+        // it does not imply privacy or anything
+
+        spec: Spec,
+        struct Spec {
+            // list-maps are ordered maps whose keys appear in their body.
+            // not specifying a key defaults to `key: name`.
+            // Go represents them as lists with specific merge keys, other
+            // languages may represent them other ways.
+            // They are automatically merged as maps in SMD.
+
+            // optional fields must list the keyword `optional` before the
+            // type.
+
+            /// list of volumes that can be mounted by containers belonging to
+            /// the pod
+            volumes: optional list-map(value: Volume),
+
+            // ...
+
+            dnsPolicy: optional(default: ClusterFirst) DNSPolicy,
+
+            // simple-maps are unordered maps from string-type values to
+            // primitive values
+            nodeSelector: optional simple-map(value: string),
+
+            // ...
+
+            @deprecated(msg: "use serviceAccountName instead")
+            serviceAccount: optional string,
+
+            // ...
+
+            readinessGates: optional list(value: ReadinessGate),
+            struct ReadinessGate { }
+
+            // create-only fields are immutable after creation (see immutability KEP)
+            containers: create-only list-map(value: Container),
+            struct Container {
+                name: string,
+                // ...
+            }
+        }
+
+        status: Status,
+
+        struct Status {
+
+        }
+    }
+
+    wrapper DNSLabel: string validates(pattern: "[a-z0-9]([-a-z0-9]*[a-z0-9])?", max-length: 64);
+
+    struct Volume {
+        name: DNSLabel,
+        source: Source,
+
+        // unions contain several variations, only one of which
+        // may be set at any given time.  By default, they have
+        // a tag of `type`.
```

Comment:
What about all the existing untagged unions?

Reply:
```
union(untagged: true) VolumeSource {
...
}
```

I think the example showing this might've been elided for brevity, but it should be explained down in the grammar notes.

Can add an example or note back in :-)

Reply:
https://github.com/kubernetes/enhancements/pull/2310/files/9a20f129f0292dc38768fb5ad140ba78fa17331f#diff-df4367b080c118ce5bcbfb8268cc95017aba2f4ff2adccb2563db13e9e8d0d7cR516

In the grammar ^, FYI